# **Ayudantía 2: Machine Learning**
**Ayudante:** Francisco Meneses Villegas (francisco.meneses@alumnos.ucn.cl)

El objetivo de esta ayudantía es que usted comprenda cómo manejar datos en proyectos de Machine Learning, desde la carga hasta el preprocesamiento, para prepararlos correctamente para su uso en modelos.

**Objetivos específicos:**
1. **Cargar distintos tipos de datos:** Explorar diversas formas de cargar datasets desde diferentes fuentes (archivos JSON, YAML y XML).
2. **Exploración de los datos:** Realizar un análisis exploratorio de los datos (EDA) para entender la distribución, correlaciones y detectar posibles outliers o anomalías.
3. **Uso de expresiones regulares:** Implementar expresiones regulares para manipular y limpiar textos en los datos.
4. **Preprocesamiento de datos:**
  - **Selección y extracción de características:** Aplicar técnicas para seleccionar las variables más relevantes y extraer nuevas características útiles.
  - **Gestión de valores nulos:** Manejar datos faltantes mediante imputación o eliminación de valores nulos.
  - **Escalado de datos:** Normalizar o estandarizar los datos para mejorar la performance de los modelos.
  - **Codificación de características:** Transformar variables categóricas en un formato adecuado para los algoritmos de Machine Learning (por ejemplo, One-Hot Encoding).

# 1. Carga de Datos

Los datos están contenidos en 3 archivos:
- games_description.json
- games_ranking.yaml
- steam_game_reviews.xml

Los cuales corresponden a un compilado de reviews y clasificaciones de los primeros 40 juegos según el ranking de steam

In [1]:
from google.colab import drive
import pandas as pd
import json
import yaml
from bs4 import BeautifulSoup

drive.mount('/gdrive')

# Ruta a los archivos en Google Drive
json_file_path = '/gdrive/MyDrive/MachineLearning/Ayudantia2/games_description.json'
yaml_file_path = '/gdrive/MyDrive/MachineLearning/Ayudantia2/games_ranking.yaml'
xml_file_path = '/gdrive/MyDrive/MachineLearning/Ayudantia2/steam_game_reviews.xml'

Drive already mounted at /gdrive; to attempt to forcibly remount, call drive.mount("/gdrive", force_remount=True).


In [2]:
# Cargar archivo JSON
with open(json_file_path, 'r') as json_file:
    json_data = json.load(json_file)
    df_json = pd.json_normalize(json_data)

# Cargar archivo YAML
with open(yaml_file_path, 'r') as yaml_file:
    yaml_data = yaml.safe_load(yaml_file)
    df_yaml = pd.json_normalize(yaml_data)

# Cargar archivo XML usando BeautifulSoup
with open(xml_file_path, 'r') as xml_file:
    soup = BeautifulSoup(xml_file, 'xml')

In [3]:
# Extraer datos del XML usando BeautifulSoup
xml_data = []
items = soup.find_all('item')

for item in items:
    item_data = {
        'review': item.find('review').text,
        'hours_played': item.find('hours_played').text,
        'helpful': item.find('helpful').text,
        'funny': item.find('funny').text,
        'recommendation': item.find('recommendation').text,
        'date': item.find('date').text,
        'game_name': item.find('game_name').text,
        'username': item.find('username').text
    }
    xml_data.append(item_data)

# Convertir los datos en un DataFrame
df_xml = pd.DataFrame(xml_data)

In [4]:
df_json.head()

name  \
0                Black Myth: Wukong   
1                  Counter-Strike 2   
2  Warhammer 40,000: Space Marine 2   
3                    Cyberpunk 2077   
4                        ELDEN RING   

                                   short_description  \
0  Black Myth: Wukong is an action RPG rooted in ...   
1  For over two decades, Counter-Strike has offer...   
2  Embody the superhuman skill and brutality of a...   
3  Cyberpunk 2077 is an open-world, action-advent...   
4  THE CRITICALLY ACCLAIMED FANTASY ACTION RPG. R...   

                                    long_description  \
0  About This Game\n\t\t\t\t\t\t\tBlack Myth: Wuk...   
1  About This Game\n\t\t\t\t\t\t\tFor over two de...   
2  About This Game\nEmbody the superhuman skill a...   
3  About This Game\nCyberpunk 2077 is an open-wor...   
4  About This Game\nTHE CRITICALLY ACCLAIMED FANT...   

                                              genres  \
0  ['Mythology', 'Action RPG', 'Action', 'RPG', '...   
1  ['FPS', 'Shooter', 'Multiplayer', 'Competitive...   
2  ['Warhammer 40K', 'Action', 'Third-Person Shoo...   
3  ['Cyberpunk', 'Open World', 'Nudity', 'RPG', '...   
4  ['Souls-like', 'Dark Fantasy', 'Open World', '...   

                          minimum_system_requirement  \
0  ['Requires a 64-bit processor and operating sy...   
1  ['OS: Windows® 10', 'Processor: 4 hardware CPU...   
2  ['Requires a 64-bit processor and operating sy...   
3  ['Requires a 64-bit processor and operating sy...   
4  ['Requires a 64-bit processor and operating sy...   

                        recommend_system_requirement  release_date  \
0  ['Requires a 64-bit processor and operating sy...  19 Aug, 2024   
1  ['OS: Windows® 10', 'Processor: 4 hardware CPU...  21 Aug, 2012   
2  ['Requires a 64-bit processor and operating sy...   9 Sep, 2024   
3  ['Requires a 64-bit processor and operating sy...  10 Dec, 2020   
4  ['Requires a 64-bit processor and operating sy...  25 Feb, 2022   

                developer                                          publisher  \
0        ['Game Science']                                   ['Game Science']   
1               ['Valve']                                          ['Valve']   
2   ['Saber Interactive']                            ['Focus Entertainment']   
3      ['CD PROJEKT RED']                                 ['CD PROJEKT RED']   
4  ['FromSoftware, Inc.']  ['FromSoftware, Inc.', 'Bandai Namco Entertain...   

     overall_player_rating number_of_reviews_from_purchased_people  \
0  Overwhelmingly Positive                               (654,820)   
1            Very Positive                             (8,313,603)   
2            Very Positive                (81% of 62,791) All Time   
3            Very Positive                               (680,264)   
4            Very Positive                               (705,261)   

  number_of_english_reviews                                               link  
0                    51,931  https://store.steampowered.com/app/2358720/Bla...  
1                 2,258,990  https://store.steampowered.com/app/730/Counter...  
2                    51,920  https://store.steampowered.com/app/2183900/War...  
3                   324,124  https://store.steampowered.com/app/1091500/Cyb...  
4                   491,741  https://store.steampowered.com/app/1245620/ELD...

In [5]:
df_yaml.head()

game_name   genre rank_type  rank
0                  Counter-Strike 2  Action     Sales     1
1  Warhammer 40,000: Space Marine 2  Action     Sales     2
2                    Cyberpunk 2077  Action     Sales     3
3                Black Myth: Wukong  Action     Sales     4
4                        ELDEN RING  Action     Sales     5

In [6]:
df_xml.head()

review hours_played helpful  \
0  The game itself is also super fun. The PvP and...         39.9   1,152   
1  Never cared much about Warhammer until this ga...         91.5     712   
2  A salute to all the fallen battle brothers who...         43.3     492   
3  this game feels like it was made in the mid 20...         16.8     661   
4  Reminds me of something I've lost. A genuine g...         24.0     557   

  funny recommendation          date                         game_name  \
0    13    Recommended  14 September  Warhammer 40,000: Space Marine 2   
1   116    Recommended  13 September  Warhammer 40,000: Space Marine 2   
2    33    Recommended  14 September  Warhammer 40,000: Space Marine 2   
3    15    Recommended  14 September  Warhammer 40,000: Space Marine 2   
4     4    Recommended  12 September  Warhammer 40,000: Space Marine 2   

                             username  
0  Sentinowl\n224 products in account  
1    userpig\n248 products in account  
2  Imparat0r\n112 products in account  
3                      Fattest_falcon  
4        Jek\n410 products in account

# 2. Análisis Exploratorio de Datos (EDA)

Necesitamos entender los dataframes que se nos presentan, analizar la distribución de sus datos, tipos de variables, correlaciones y posibles anomalías

## df_json

In [7]:
df_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 290 entries, 0 to 289
Data columns (total 13 columns):
 #   Column                                   Non-Null Count  Dtype 
---  ------                                   --------------  ----- 
 0   name                                     290 non-null    object
 1   short_description                        277 non-null    object
 2   long_description                         290 non-null    object
 3   genres                                   290 non-null    object
 4   minimum_system_requirement               290 non-null    object
 5   recommend_system_requirement             290 non-null    object
 6   release_date                             290 non-null    object
 7   developer                                290 non-null    object
 8   publisher                                290 non-null    object
 9   overall_player_rating                    290 non-null    object
 10  number_of_reviews_from_purchased_people  290 non-null    objec

Notamos que en el caso anterior, el dataframe se compone de 13 columnas, pero que todas aparecen como un dato del tipo 'Object', lo cual no es del todo correcto, ya que la release_date debería ser de tipo 'date' y number_of_reviews_from_purchased_people junto a number_of_english_reviews deberían ser de tipo numérico

Notamos además que en la columna short_description hay solo 277 elementos no nulos de un total de 290, lo que significa que habrán elementos nulos en dicha columna

## df_yaml

In [8]:
df_yaml.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 672 entries, 0 to 671
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   game_name  672 non-null    object
 1   genre      672 non-null    object
 2   rank_type  672 non-null    object
 3   rank       672 non-null    int64 
dtypes: int64(1), object(3)
memory usage: 21.1+ KB


Con el análisis simple de información no notamos problemas en el dataframe df_yaml

## df_xml

In [9]:
df_xml.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 992153 entries, 0 to 992152
Data columns (total 8 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   review          992153 non-null  object
 1   hours_played    992153 non-null  object
 2   helpful         992153 non-null  object
 3   funny           992153 non-null  object
 4   recommendation  992153 non-null  object
 5   date            992153 non-null  object
 6   game_name       992153 non-null  object
 7   username        992153 non-null  object
dtypes: object(8)
memory usage: 60.6+ MB


Vemos que nuevamente todas las columnas tiene datos de tipo 'Object', cuando hours_played debería ser de tipo float y helpful junto a funny deberían ser de tipo int, además la columna date debería ser de tipo datetime

# 3. Limpieza con Expresiones Regulares

## Dataframe con JSON

Primero notamos que hay una serie de columnas que tiene corchetes y comillas separando los elementos dentro de las mismas columnas, en función de la limpieza, eliminaremos dichos corchetes y comillas

In [10]:
import re

# Función para eliminar corchetes y comillas con expresiones regulares
def clean_text_using_regex(value):
    if isinstance(value, str):
        return re.sub(r"[\[\]\'\"]", "", value)
    return value

columns_to_clean = ['developer', 'publisher', 'genres', 'minimum_system_requirement', 'recommend_system_requirement']

for column in columns_to_clean:
    df_json[column] = df_json[column].apply(clean_text_using_regex)

Notamos de igual forma que la columna _**long_description**_ hay elementos que contienen "About This Game\n\t\t\t\t\t\t\t" al inicio, dichos problemas de formateo también podemos solucionarlos con expresiones regulares

In [11]:
# Función para eliminar el texto "About This Game" seguido de saltos de línea y tabulaciones al inicio de la descripción
def clean_long_description(value):
    if isinstance(value, str):
        return re.sub(r"^About This Game\s*\n*\t*", "", value)
    return value

df_json['long_description'] = df_json['long_description'].apply(clean_long_description)

Sabemos que en la columna _**number_of_english_reviews**_ debería ir un valor de tipo int así que lo parseamos

In [12]:
# Función para eliminar comas y convertir a entero
def convert_to_int(value):
    if isinstance(value, str):
        value = re.sub(r",", "", value)
        return int(value)  # Convertir a entero
    return value

# Aplicar la función a la columna 'number_of_english_reviews'
df_json['number_of_english_reviews'] = df_json['number_of_english_reviews'].apply(convert_to_int)

Finalmente en la columna _**number_of_reviews_from_purchased_people**_ hay datos de distinto elementos, estos los podemos parsear

In [13]:
# Función para manejar ambos tipos de formatos y limpiar caracteres adicionales
def parse_reviews(value):
    if isinstance(value, str):
        # Caso 1: (81% of 62,791)
        match = re.search(r"of\s([\d,]+)", value)
        if match:
            return int(re.sub(r",", "", match.group(1)))

        # Caso 2: (654,820) y eliminar cualquier texto adicional
        cleaned_value = re.sub(r"[^\d,]", "", value)

        if cleaned_value:  # Verificar que el valor no esté vacío después de la limpieza
            return int(re.sub(r",", "", cleaned_value))

    return None

df_json['number_of_reviews_from_purchased_people'] = df_json['number_of_reviews_from_purchased_people'].apply(parse_reviews)

In [14]:
df_json.dtypes

name                                        object
short_description                           object
long_description                            object
genres                                      object
minimum_system_requirement                  object
recommend_system_requirement                object
release_date                                object
developer                                   object
publisher                                   object
overall_player_rating                       object
number_of_reviews_from_purchased_people    float64
number_of_english_reviews                    int64
link                                        object
dtype: object

In [15]:
# Rellenar valores NaN con 0 antes de convertir a int
df_json['number_of_reviews_from_purchased_people'] = df_json['number_of_reviews_from_purchased_people'].fillna(0).astype('int64')

In [16]:
df_json.tail()

name  \
285                                      MXGP PRO   
286  MXGP 2019 - The Official Motocross Videogame   
287                                        Ride 2   
288                                 Tanuki Sunset   
289         Grand Mountain Adventure: Wonderlands   

                                     short_description  \
285  Adjust your bike's configuration like a mechan...   
286  Experience the excitement of the 2019 season f...   
287  The fastest and most iconic bikes in the world...   
288  Master the longboard as you drift through stun...   
289  Grand Mountain Adventure: Wonderlands lets you...   

                                      long_description  \
285  The official video game of the Motocross Champ...   
286  The official Motocross Championship video game...   
287  Welcome to the temple of motorcycles, the only...   
288  Grab Your Longboard and Race to the TopMeet Ta...   
289  Grand Mountain Adventure: Wonderlands lets you...   

                                                genres  \
285  Simulation, Racing, Sports, Motocross, Motorbi...   
286  Simulation, Racing, Sports, Motorbike, Motocro...   
287  Racing, Sports, Simulation, Motorbike, Bikes, ...   
288  Racing, Action, Indie, Sports, Adventure, Skat...   
289  Simulation, Adventure, Sports, Casual, Indie, ...   

                            minimum_system_requirement  \
285  Requires a 64-bit processor and operating syst...   
286  Requires a 64-bit processor and operating syst...   
287  OS *: Windows® 7 SP1 / Windows® 8 / Windows® 8...   
288  OS *: Windows 7 (32-bit), Processor: Dual Core...   
289  Requires a 64-bit processor and operating syst...   

                          recommend_system_requirement  release_date  \
285  Requires a 64-bit processor and operating syst...  29 Jun, 2018   
286  Requires a 64-bit processor and operating syst...  27 Aug, 2019   
287  OS *: Windows® 7 SP1 64-Bit / Windows® 8 64-Bi...   7 Oct, 2016   
288  OS *: Windows 7 (32-bit), Processor: Dual Core...   4 Dec, 2020   
289  Requires a 64-bit processor and operating syst...  10 Mar, 2022   

            developer         publisher overall_player_rating  \
285  Milestone S.r.l.  Milestone S.r.l.         Very Positive   
286  Milestone S.r.l.  Milestone S.r.l.         Very Positive   
287  Milestone S.r.l.  Milestone S.r.l.         Very Positive   
288      Rewind Games      Rewind Games         Very Positive   
289       Toppluva AB          Microids         Very Positive   

     number_of_reviews_from_purchased_people  number_of_english_reviews  \
285                                      777                        433   
286                                      858                        446   
287                                     2335                       1045   
288                                      420                        451   
289                                      196                        246   

                                                  link  
285  https://store.steampowered.com/app/798290/MXGP...  
286  https://store.steampowered.com/app/1018160/MXG...  
287  https://store.steampowered.com/app/477770/Ride...  
288  https://store.steampowered.com/app/1251460/Tan...  
289  https://store.steampowered.com/app/1247360/Gra...

## Dataframe con XML

En la columna _**review**_ del dataframe de XML podrían quedar etiquetas propias del HTML que no hayan sido eliminadas o parseadas correctamente

In [17]:
# Reemplazar entidades HTML por sus correspondientes caracteres
df_xml['review'] = df_xml['review'].apply(lambda x: re.sub(r'&#8212;', '-', x))  # Reemplaza guiones largos por '-'
df_xml['review'] = df_xml['review'].apply(lambda x: re.sub(r'[^a-zA-Z0-9\s.,!?]', '', x))  # Eliminar caracteres no alfanuméricos

En las columnas _**helpful**_ y _**funny**_ pueden haber valores separados por comas ( , ) por lo que los cambiaremos a valores del tipo numérico

In [18]:
df_xml['helpful'] = df_xml['helpful'].apply(lambda x: re.sub(r',', '', x))
df_xml['helpful'] = pd.to_numeric(df_xml['helpful'], errors='coerce')  # Convertir a tipo numérico

df_xml['funny'] = df_xml['funny'].apply(lambda x: re.sub(r',', '', x))
df_xml['funny'] = pd.to_numeric(df_xml['funny'], errors='coerce')

Observamos que en la columna _**username**_ hay filas en donde aparece el nombre de usuario, seguido por un salto de línea _**\n**_  y luego por la cantidad de productos asociados a la cuenta de dicho usuario, podemos separar estos elementos y agregar la columna _**products in account**_ que tenga el valor numérico que aparece en username, en caso que solo aparezca el username, la columna products in account queda con valor "1"

In [19]:
# Crear una nueva columna 'products_in_account' que almacene el número de productos
df_xml['products_in_account'] = df_xml['username'].apply(lambda x: re.findall(r'(\d+) products in account', x))

# Si se encuentra un valor, convertirlo a entero, de lo contrario asignar '1'
df_xml['products_in_account'] = df_xml['products_in_account'].apply(lambda x: int(x[0]) if x else 1)

# Limpiar la columna 'username'
df_xml['username'] = df_xml['username'].apply(lambda x: re.sub(r'\n\d+ products in account', '', x).strip())

In [20]:
df_xml.dtypes

review                 object
hours_played           object
helpful                 int64
funny                   int64
recommendation         object
date                   object
game_name              object
username               object
products_in_account     int64
dtype: object

Finalmente notamos que la columna _**hours_played**_ es de tipo object, por lo que la parsearemos a tipo float64

In [21]:
df_xml['hours_played'] = pd.to_numeric(df_xml['hours_played'], errors='coerce')

In [22]:
df_xml.dtypes

review                  object
hours_played           float64
helpful                  int64
funny                    int64
recommendation          object
date                    object
game_name               object
username                object
products_in_account      int64
dtype: object

In [23]:
df_xml.head()

review  hours_played  helpful  \
0  The game itself is also super fun. The PvP and...          39.9     1152   
1  Never cared much about Warhammer until this ga...          91.5      712   
2  A salute to all the fallen battle brothers who...          43.3      492   
3  this game feels like it was made in the mid 20...          16.8      661   
4  Reminds me of something Ive lost. A genuine ga...          24.0      557   

   funny recommendation          date                         game_name  \
0     13    Recommended  14 September  Warhammer 40,000: Space Marine 2   
1    116    Recommended  13 September  Warhammer 40,000: Space Marine 2   
2     33    Recommended  14 September  Warhammer 40,000: Space Marine 2   
3     15    Recommended  14 September  Warhammer 40,000: Space Marine 2   
4      4    Recommended  12 September  Warhammer 40,000: Space Marine 2   

         username  products_in_account  
0       Sentinowl                  224  
1         userpig                  248  
2       Imparat0r                  112  
3  Fattest_falcon                    1  
4             Jek                  410

# 4. Preprocesamiento de Datos